handlers/utils.go

```diff
@@ -22,13 +22,65 @@ const defaultPatchType = "merge"
 
 var defaultPromClientSupplier = prometheus.NewClient
 
-func checkNamespaceAccess(ctx context.Context, nsServ business.NamespaceService, namespace string) (*models.Namespace, error) {
-	return nsServ.GetNamespace(ctx, namespace)
+func checkNamespaceAccess(ctx context.Context, nsServ business.NamespaceService, namespace string, cluster string) (*models.Namespace, error) {
+	return nsServ.GetClusterNamespace(ctx, namespace, cluster)
 }
 
-func createMetricsServiceForNamespace(w http.ResponseWriter, r *http.Request, promSupplier promClientSupplier, namespace string) (*business.MetricsService, *models.Namespace) {
-	metrics, infoMap := createMetricsServiceForNamespaces(w, r, promSupplier, []string{namespace})
-	if result, ok := infoMap[namespace]; ok {
+// createMetricsServiceForNamespaceMC is used when the service will query across all clusters for the namespace.
+// It will return an error if the user does not have access to the namespace on all of the clusters.
+func createMetricsServiceForNamespaceMC(w http.ResponseWriter, r *http.Request, promSupplier promClientSupplier, nsName string) (*business.MetricsService, []models.Namespace) {
+	layer, err := getBusiness(r)
+	if err != nil {
+		RespondWithError(w, http.StatusInternalServerError, err.Error())
+		return nil, nil
+	}
+	prom, err := promSupplier()
+	if err != nil {
+		log.Error(err)
+		RespondWithError(w, http.StatusServiceUnavailable, "Prometheus client error: "+err.Error())
+		return nil, nil
+	}
+	var nsInfo []models.Namespace
+
+	nsClusters, err := layer.Namespace.GetNamespaceClusters(r.Context(), nsName)
+	if err != nil || len(nsClusters) == 0 {
+		log.Infof("No clusters found.")
+		//TODO: We cant find the cluster list, we assume it is local
+		nsClusters = make([]models.Namespace, 1)
+		nsClusters[0] = models.Namespace{Cluster: config.Get().KubernetesConfig.ClusterName, Name: nsName}
+	}
+
+	for _, nsCluster := range nsClusters {
+		_, err := checkNamespaceAccess(r.Context(), layer.Namespace, nsCluster.Name, nsCluster.Cluster)
+		// If there is no access to one of the namespace, return the error
+		if err != nil {
+			RespondWithError(w, http.StatusForbidden, "Cannot access namespace data: "+err.Error())
+			return nil, nil
+		}
+		nsInfo = append(nsInfo, nsCluster)
+
+	}
+	metrics := business.NewMetricsService(prom)
+
+	return metrics, nsInfo
+}
+
+// Get namespespace with oldest creation time
+// Used to choose between one namespace from all the clusters and use the creation timestamp for the metrics service
+func GetNsWithOldestCreationDate(namespaces []models.Namespace) *models.Namespace {
```

Comment:
```suggestion
// GetOldestNamespace is a convenience function that takes a list of Namespaces and returns the
// Namespace with the oldest CreationTimestamp.  In a tie, preference is towards the head of the list.
func GetOldestNamespace(namespaces []models.Namespace) *models.Namespace {
```

Function comments in Go typically start with the function name.  In this case I suggest we also shorten the name and remove the abbreviation, but you can decide what you like best.

Reply:
Updated